File: src/main/scala/com/pygostylia/osprey/MinecraftInputStream.scala
package com.pygostylia.osprey

import java.io.{ByteArrayInputStream, DataInputStream, InputStream}

class MinecraftInputStream(is: InputStream) extends DataInputStream(is) {
  def this(bs: Array[Byte]) = this(new ByteArrayInputStream(bs))

  def readVarInt(): Int = {
    var i = 0
    var n = 0
    while (true) {
      val b = read()
      if (0 to 127 contains b) {
        i |= b << (7 * n)
        return i
      }
      i |= (b & 0x7f) << (7 * n)
      n += 1
    }
    i
  }

  def readString(): String = {
    val len = readVarInt()
    val bytes = readNBytes(len)
    new String(bytes)
  }

  def readLocation(): BlockPosition = {
    new BlockPosition(readLong())
  }

  def readPosition(): EntityPosition = {
    val x = readDouble()
    val y = readDouble()
    val z = readDouble()
    val yaw = readFloat()
    val pitch = readFloat()
    new EntityPosition(x, y, z, yaw, pitch)
  }
}

File: src/main/java/com/pygostylia/osprey/MinecraftInputStream.java
package com.pygostylia.osprey;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;

public class MinecraftInputStream extends DataInputStream {
    public int type;
    public int originalLen;
    public int compressedLen;

    public MinecraftInputStream(byte[] buf, int len) throws IOException {
        super(new ByteArrayInputStream(buf));
        originalLen = len;
    }

    public MinecraftInputStream(InputStream is) throws IOException {
        super(is);
    }

    public int readVarInt() throws IOException {
        return VarInt.read(this);
    }

    public String readString() throws IOException {
        int length = readVarInt();
        byte[] data = new byte[length];
        read(data);
        return new String(data);
    }

    public BlockPosition readLocation() throws IOException {
        return new BlockPosition(Protocol.readLong(this));
    }

    public EntityPosition readPosition() throws IOException {
        var x = readDouble();
        var y = readDouble();
        var z = readDouble();
        var yaw = readFloat();
        var pitch = readFloat();
        return new EntityPosition(x, y, z, yaw, pitch);
    }
}
